fixes for building on armv7

noblige opened the pull request:
Compiles and runs on target.
https://github.com/Project-OSRM/osrm-backend/wiki/Building-for-ARMv7

@danpat

TheMarex commented on include/util/compat.hpp:
```diff
@@ -0,0 +1,36 @@
+
+namespace std
+{
+    template <typename T> std::string to_string(const T& n)
```
Add comment:
```
See discussion https://answers.launchpad.net/gcc-arm-embedded/+question/238327
```

TheMarex commented on include/util/compat.hpp:
```diff
@@ -0,0 +1,36 @@
+        return stm.str();
+    }
+
+    inline double round(double d)
```
Add comment.
```
These might be fixed with a more recent gcc. See discussion on w.r.t. to ChromeOS gcc:
https://code.google.com/p/android/issues/detail?id=54418
```

TheMarex commented on CMakeLists.txt:
```diff
@@ -322,27 +335,29 @@ target_link_libraries(osrm-extract osrm_extract ${Boost_LIBRARIES})
-    ${TBB_LIBRARIES}
-    ${MAYBE_RT_LIBRARY}
-    ${MAYBE_COVERAGE_LIBRARIES})
+if(NOT ROUTED_STANDALONE)
```
If we really want to support this, we need to split the CMakeList not add a bunch of switches all over the place.

@daniel-j-h you did this for the `unit_tests` what are you experiences w.r.t to complexity here?

daniel-j-h replied:
The unit tests are currently doing this: provide a recursive CMakeLists.txt and register it at the end of the top level CMakeLists.txt. This way, we only require the Boost Test library when unit tests are requested.

TheMarex commented on CMakeLists.txt:
```diff
@@ -245,9 +248,13 @@ endif()
+  add_dependency_includes(${OSMIUM_INCLUDE_DIR})
+else()
+  # transitive dependency
+  include_directories(SYSTEM ${OSMIUM_INCLUDE_DIR})
```
why do you have a switch here? Adding OSMIUM include path to the search path while not calling find_package will be a no-op.

daniel-j-h commented on CMakeLists.txt:
```diff
@@ -215,6 +215,9 @@ if("${LINKER_VERSION}" MATCHES "GNU gold" OR "${LINKER_VERSION}" MATCHES "GNU ld
 endif ()
+if (STATIC_LIBSTDCXX)
+  set(LINKER_FLAGS "${LINKER_FLAGS} -static-libstdc++")
+endif()
```
CMake picks up your environment's `CXXFLAGS` and `LDFLAGS` --- there should be no need for this switch here, just export the appropriate flags.

noblige commented:
Addressed comments except for `CMakeList` refactoring, which feels like a bigger task than I'm comfortable with :) I wonder if we can do this as the separate next step?

TheMarex commented:
@noblige in that case please remove the flag to build `osrm-routed` standalone. We can fix this in another iteration. :+1:

noblige commented:
@TheMarex reverted `OSRM_STANDALONE` change - good to go now

oxidase commented on include/util/compat.hpp:
```diff
@@ -0,0 +1,39 @@
+#ifndef OSRM_COMPAT_HPP
+#define OSRM_COMPAT_HPP
+
+#if defined(__ARM_ARCH_7A__)
```
Fixes are not really related to ARM architecture but to a specific libstdc++ version of the used toolchain. I would suggest to make it as a patch and apply it before building with the specific toolchain. Otherwise the patch brings redefinition errors for newer libstc++
```
[  1%] Building CXX object CMakeFiles/ENGINE.dir/src/engine/polyline_compressor.cpp.o
In file included from /home/miha/foss/osrm-backend/src/engine/polyline_compressor.cpp:3:0:
/home/miha/foss/osrm-backend/include/util/compat.hpp: In function ‘double std::round(double)’:
/home/miha/foss/osrm-backend/include/util/compat.hpp:21:33: error: ‘double std::round(double)’ conflicts with a previous declaration
     inline double round(double d)
                                 ^
In file included from /usr/arm-linux-gnueabihf/include/math.h:83:0,
                 from /usr/arm-linux-gnueabihf/include/c++/5/cmath:44,
```

noblige replied:
@oxidase thanks! I will close this for now
